In [1]:
# Import libraries
from dotenv import load_dotenv
import os
import psycopg2
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import ipywidgets as widgets
from IPython.display import display

In [2]:
# Load environment variables
load_dotenv()
db_user = os.getenv('DB_USER')
db_password = os.getenv('DB_PASSWORD')
db_host = os.getenv('DB_HOST')
db_name = os.getenv('DB_NAME')

In [3]:
# Connect to the Stack Overflow database
conn = psycopg2.connect(
    host=db_host,
    database=db_name,
    user=db_user,
    password=db_password
)

In [4]:
# Define HEART metrics for Stack Overflow
heart_metrics = {
    'Happiness': 'Measure of user satisfaction and happiness with the Questions & Answers product.',
    'Engagement': 'Measure of user interaction and activity on the platform.',
    'Adoption': 'Measure of user acquisition and growth of Stack Overflow user base.',
    'Retention': 'Measure of user retention and continued usage of the platform over time.',
    'Task Success': 'Measure of user ability to accomplish their goals and tasks effectively on Stack Overflow.'
}

When moving through the HEART framework's "Goals Signals Metrics" process, it can be useful to lay out these things out in a table like the following:

|  | Goals | Signals | Metrics |
| --- | --- | --- | --- |
| Happiness | Users find the product helpful, fun and easy to use | This could be signalled by users recommending Stack Overflow Questions & Answers to others.  | The score number being above a 164. |
| Engagement | Users enjoy product content and keep engaging with it | This could be signalled by the amount of time spent on the site, or the number of interactions with the content (comments, posts, voting, sharing). | total total interactions with the content (sum of comments, posts, shares, votes) |
| Adoption | New users see the value in the product | This could be signalled by the number of new user registrations. | total number of new user registrations  |
| Retention | Users keep coming back to the app to complete a key action | This could be signalled by the user retention rate over time. | percentage of users retained over specific time periods. |
| Task Success | Users complete their goal quickly and easily | This could be signalled by task success rate. | percentage of users successfully completing their tasks |

# Build SQL Queries

Write SQL queries to calculate the metrics

In [5]:
# Happiness

with conn.cursor() as cursor:
    cursor.execute("SELECT COUNT(*) AS good_post_scores FROM posts WHERE score > 164;")
    df = pd.DataFrame(cursor.fetchall(), columns=[desc[0] for desc in cursor.description])

print(df)

   good_post_scores
0             84548


In [6]:
with conn.cursor() as cursor:
    cursor.execute("SELECT SUM(comment_count) AS total_comments FROM posts;")
    df = pd.DataFrame(cursor.fetchall(), columns=['Total comments'])

print(df)

   Total comments
0        87480313


In [7]:
# Adoption

with conn.cursor() as cursor:
    cursor.execute("""
        WITH current_year_data AS (
            SELECT COUNT(*) AS new_user_registrations 
            FROM users 
            WHERE EXTRACT(YEAR FROM creation_date) = EXTRACT(YEAR FROM CURRENT_DATE)
        ),
        previous_year_data AS (
            SELECT COUNT(*) AS new_user_registrations 
            FROM users 
            WHERE EXTRACT(YEAR FROM creation_date) = EXTRACT(YEAR FROM CURRENT_DATE) - 1
            AND EXTRACT(MONTH FROM creation_date) BETWEEN 1 AND 6
        ) 
        SELECT 
            (SELECT new_user_registrations FROM current_year_data) AS current_year_registrations, 
            (SELECT new_user_registrations FROM previous_year_data) AS previous_year_registrations;
    """)
    result = cursor.fetchall()
df = pd.DataFrame(result, columns=['Current Year Registrations (Jan-June)', 'Previous Year Registrations (Jan-June)'])
display(df)

Current Year Registrations (Jan-June)  \
0                                 428090   

   Previous Year Registrations (Jan-June)  
0                                 1627247

In [8]:
# Retention

with conn.cursor() as cursor:
    cursor.execute("""
        WITH user_counts AS (
            SELECT
                DATE_TRUNC('month', creation_date) AS account_creation_month,
                COUNT(*) AS num_new_users
            FROM users
            GROUP BY account_creation_month
        ),
        retained_users AS (
            SELECT
                DATE_TRUNC('month', u.creation_date) AS account_creation_month,
                COUNT(*) AS num_retained_users
            FROM users u
            WHERE EXISTS (
                SELECT 1
                FROM users u2
                WHERE u2.account_id = u.account_id
                AND DATE_TRUNC('month', u2.last_access_date) >= DATE_TRUNC('month', u.creation_date) + INTERVAL '1 month'
                AND DATE_TRUNC('month', u2.last_access_date) <= DATE_TRUNC('month', u.creation_date) + INTERVAL '6 month'
            )
            GROUP BY account_creation_month
        )
        SELECT
            uc.account_creation_month,
            uc.num_new_users,
            COALESCE(num_retained_users, 0) AS num_retained_users
        FROM user_counts uc
        LEFT JOIN retained_users ru ON uc.account_creation_month = ru.account_creation_month
        ORDER BY uc.account_creation_month;
    """)

    # Fetch the result
    result = cursor.fetchall()

# Create a DataFrame from the query result
df = pd.DataFrame(result, columns=['Account Creation Month', 'New Users', 'Retained Users'])

# Display the DataFrame as a table in the Jupyter notebook
display(df)

Account Creation Month  New Users  Retained Users
0               2008-07-01          9               1
1               2008-08-01       2806              65
2               2008-09-01       9643             595
3               2008-10-01       3544             275
4               2008-11-01       2857             259
..                     ...        ...             ...
172             2022-11-01     264690          123076
173             2022-12-01     249281          102843
174             2023-01-01     217014           92081
175             2023-02-01     184360           44613
176             2023-03-01      26716               0

[177 rows x 3 columns]

In [9]:
# Task Success

with conn.cursor() as cursor:
    cursor.execute("""
        SELECT
            DATE_TRUNC('month', p.creation_date) AS "Month",
            COUNT(DISTINCT CASE WHEN p.accepted_answer_id IS NOT NULL THEN p.id END) AS "Accepted Answers",
            COUNT(DISTINCT CASE WHEN p.post_type_id = 1 THEN p.id END) AS "Total Questions",
            COUNT(DISTINCT CASE WHEN p.accepted_answer_id IS NOT NULL THEN p.id END) /
                NULLIF(COUNT(DISTINCT CASE WHEN p.post_type_id = 1 THEN p.id END), 0) AS "Task Success Rate"
        FROM posts p
        WHERE p.creation_date >= DATE_TRUNC('year', CURRENT_DATE) - INTERVAL '2 year'
        GROUP BY DATE_TRUNC('month', p.creation_date)
        ORDER BY DATE_TRUNC('month', p.creation_date);
    """)

    # Fetch the result
    result = cursor.fetchall()

# Create a DataFrame from the query result
df = pd.DataFrame(result, columns=['Month', 'Accepted Answers', 'Total Questions', 'Task Success Rate'])

# Display the DataFrame as a table in the Jupyter notebook
display(df)

Month  Accepted Answers  Total Questions  Task Success Rate
0  2021-01-01             62838           141177                  0
1  2021-02-01             58783           132950                  0
2  2021-03-01             65181           150434                  0
3  2021-04-01             58948           137644                  0
4  2021-05-01             57895           135360                  0
5  2021-06-01             55057           130545                  0
6  2021-07-01             53116           125461                  0
7  2021-08-01             52481           123714                  0
8  2021-09-01             50519           121266                  0
9  2021-10-01             50524           120532                  0
10 2021-11-01             49626           120981                  0
11 2021-12-01             46176           133566                  0
12 2022-01-01             49592           142958                  0
13 2022-02-01             45800           139896                  0
14 2022-03-01             48549           152548                  0
15 2022-04-01             44866           141310                  0
16 2022-05-01             44642           144693                  0
17 2022-06-01             43066           139296                  0
18 2022-07-01             43309           137909                  0
19 2022-08-01             44213           141097                  0
20 2022-09-01             40306           131418                  0
21 2022-10-01             40350           135401                  0
22 2022-11-01             38266           158425                  0
23 2022-12-01              3159            20538                  0

The above indicates that there are 84,548 posts with a score above 164. Having a score above 164 is considered good. The score of a post is calculated by:
   * number of upvotes - number of downvotes


In [10]:
# Define SQL queries
sql_queries = {
    'Happiness': "SELECT COUNT(*) AS good_post_scores FROM posts WHERE score > 164;",
    'Engagement': "SELECT SUM(comment_count) AS total_comments FROM posts;",
    'Adoption': """
        WITH current_year_data AS (
            SELECT COUNT(*) AS new_user_registrations 
            FROM users 
            WHERE EXTRACT(YEAR FROM creation_date) = EXTRACT(YEAR FROM CURRENT_DATE)
        ),
        previous_year_data AS (
            SELECT COUNT(*) AS new_user_registrations 
            FROM users 
            WHERE EXTRACT(YEAR FROM creation_date) = EXTRACT(YEAR FROM CURRENT_DATE) - 1
            AND EXTRACT(MONTH FROM creation_date) BETWEEN 1 AND 6
        ) 
        SELECT 
            (SELECT new_user_registrations FROM current_year_data) AS current_year_registrations, 
            (SELECT new_user_registrations FROM previous_year_data) AS previous_year_registrations;
    """,
    'Retention': """
        WITH user_counts AS (
            SELECT
                DATE_TRUNC('month', creation_date) AS account_creation_month,
                COUNT(*) AS num_new_users
            FROM users
            GROUP BY account_creation_month
        ),
        retained_users AS (
            SELECT
                DATE_TRUNC('month', u.creation_date) AS account_creation_month,
                COUNT(*) AS num_retained_users
            FROM users u
            WHERE EXISTS (
                SELECT 1
                FROM users u2
                WHERE u2.account_id = u.account_id
                AND DATE_TRUNC('month', u2.last_access_date) >= DATE_TRUNC('month', u.creation_date) + INTERVAL '1 month'
                AND DATE_TRUNC('month', u2.last_access_date) <= DATE_TRUNC('month', u.creation_date) + INTERVAL '6 month'
            )
            GROUP BY account_creation_month
        )
        SELECT
            uc.account_creation_month,
            uc.num_new_users,
            COALESCE(num_retained_users, 0) AS num_retained_users
        FROM user_counts uc
        LEFT JOIN retained_users ru ON uc.account_creation_month = ru.account_creation_month
        ORDER BY uc.account_creation_month;
    """,
    'Task_Success': """
        SELECT
            DATE_TRUNC('month', p.creation_date) AS "Month",
            COUNT(DISTINCT CASE WHEN p.accepted_answer_id IS NOT NULL THEN p.id END) AS "Accepted Answers",
            COUNT(DISTINCT CASE WHEN p.post_type_id = 1 THEN p.id END) AS "Total Questions",
            COUNT(DISTINCT CASE WHEN p.accepted_answer_id IS NOT NULL THEN p.id END) /
                NULLIF(COUNT(DISTINCT CASE WHEN p.post_type_id = 1 THEN p.id END), 0) AS "Task Success Rate"
        FROM posts p
        WHERE p.creation_date >= DATE_TRUNC('year', CURRENT_DATE) - INTERVAL '2 year'
        GROUP BY DATE_TRUNC('month', p.creation_date)
        ORDER BY DATE_TRUNC('month', p.creation_date);
    """
}

# Execute SQL queries and display results as tables
for metric, query in sql_queries.items():
    df = pd.read_sql_query(query, conn)
    print(f"--- {metric} ---")
    print(df)
    print("\n")

/var/folders/7g/8krtxjqx2s1d47kw7kbhb2h80000gp/T/ipykernel_40942/1692914169.py:67: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, conn)


--- Happiness ---
   good_post_scores
0             84548




/var/folders/7g/8krtxjqx2s1d47kw7kbhb2h80000gp/T/ipykernel_40942/1692914169.py:67: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, conn)
